src/ObjectPool/src/DefaultObjectPool.cs

```diff
@@ -39,66 +38,62 @@ public DefaultObjectPool(IPooledObjectPolicy<T> policy)
-        _isDefaultPolicy = IsDefaultPolicy();
-
-        // -1 due to _firstItem
-        _items = new ObjectWrapper[maximumRetained - 1];
-
-        bool IsDefaultPolicy()
-        {
-            var type = policy.GetType();
-
-            return type.IsGenericType && type.GetGenericTypeDefinition() == typeof(DefaultPooledObjectPolicy<>);
-        }
+        // cache the target interface methods, to avoid interface lookup overhead
+        _createFunc = policy.Create;
+        _returnFunc = policy.Return;
+        _maxCapacity = maximumRetained - 1;  // -1 to account for _fastItem
     }
 
     /// <inheritdoc />
     public override T Get()
     {
-        var item = _firstItem;
-        if (item == null || Interlocked.CompareExchange(ref _firstItem, null, item) != item)
+        var item = _fastItem;
+        if (item == null || Interlocked.CompareExchange(ref _fastItem, null, item) != item)
         {
-            var items = _items;
-            for (var i = 0; i < items.Length; i++)
+            if (_items.TryDequeue(out item))
             {
-                item = items[i].Element;
-                if (item != null && Interlocked.CompareExchange(ref items[i].Element, null, item) == item)
-                {
-                    return item;
-                }
+                _ = Interlocked.Decrement(ref _numItems);
+                return item;
             }
 
-            item = Create();
+            // no object available, so go get a brand new one
+            return _createFunc();
         }
 
         return item;
     }
 
-    // Non-inline to improve its code quality as uncommon path
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private T Create() => _fastPolicy?.Create() ?? _policy.Create();
-
     /// <inheritdoc />
     public override void Return(T obj)
     {
-        if (_isDefaultPolicy || (_fastPolicy?.Return(obj) ?? _policy.Return(obj)))
+        _ = ReturnCore(obj);
+    }
+
+    /// <summary>
+    /// Returns an object to the pool.
+    /// </summary>
+    /// <returns>true if the object was returned to the pool</returns>
+    private protected bool ReturnCore(T obj)
+    {
+        if (!_returnFunc(obj))
+        {
+            // policy says to drop this object
+            return false;
+        }
+
+        if (_fastItem != null || Interlocked.CompareExchange(ref _fastItem, obj, null) != null)
         {
-            if (_firstItem != null || Interlocked.CompareExchange(ref _firstItem, obj, null) != null)
+            if (Interlocked.Increment(ref _numItems) <= _maxCapacity)
             {
-                var items = _items;
-                for (var i = 0; i < items.Length && Interlocked.CompareExchange(ref items[i].Element, obj, null) != null; ++i)
-                {
-                }
+                _items.Enqueue(obj);
+                return true;
             }
+
+            // no room, clean up the count and drop the object on the floor
+            _ = Interlocked.Decrement(ref _numItems);
```

Comment:
```suggestion
            Interlocked.Decrement(ref _numItems);
```

Reply:
Why remove the _ = sequences? This is a rule we have enabled in our code base as it serves to notify the reader that a value is being ignored. Is this frowned upon in this code base?